tooling/metamodel-generator/src/main/java/org/hibernate/processor/validation/ProcessorSessionFactory.java

```diff
@@ -196,28 +207,79 @@ private static JdbcType enumJdbcType(Element member) {
 				: IntegerJdbcType.INSTANCE;
 	}
 
-	final Map<String, Set<String>> result = new HashMap<>();
+	final Set<TypeElement> unprocessedTypes = new HashSet<>();
+	final Set<Name> visitedTypes = new HashSet<>();
+	final Map<String, Set<String>> allowedEnumLiteralsToEnumClassName = new HashMap<>();
 
 	@Override
-	Map<String, Set<String>> getAllowedEnumLiteralTexts() {
-		//TODO: elementUtil.getAllModuleElements();
-		if ( result.isEmpty() ) {
-			for (Element mod : elementUtil.getModuleElement("").getEnclosedElements()) {
-				for (Element element : mod.getEnclosedElements()) {
-					if (element.getKind() == ElementKind.ENUM) {
-						TypeElement typeElement = (TypeElement) element;
-						for (Element member : element.getEnclosedElements()) {
-							if (member.getKind() == ElementKind.ENUM_CONSTANT) {
-								String name = member.getSimpleName().toString();
-								result.computeIfAbsent( name, s -> new HashSet<>() )
-										.add( typeElement.getQualifiedName().toString() );
-							}
-						}
-					}
-				}
+	Map<String, Set<String>> getAllowedEnumLiteralsToEnumClassName() {
+		if ( !unprocessedTypes.isEmpty() ) {
+			for ( TypeElement typeElement : unprocessedTypes ) {
+				discoverEnums( typeElement, visitedTypes );
+			}
+			unprocessedTypes.clear();
+		}
+		return allowedEnumLiteralsToEnumClassName;
+	}
+
+	private void discoverEnums(TypeElement element, Set<Name> visitedTypes) {
+		if ( element == null || !visitedTypes.add( element.getQualifiedName() ) ) {
+			return;
+		}
+		final String qualifiedName = element.getQualifiedName().toString();
+		if ( element.getKind() == ElementKind.ENUM ) {
+			processEnum( element );
+		}
+		else {
+			Metamodel metaEntity = context.getMetaEntity( qualifiedName );
+			if ( metaEntity != null ) {
+				discoverEnums( metaEntity, visitedTypes );
+			}
+			else if ( hasAnnotation( element, ENTITY ) ) {
+				context.addMetaEntity(
+						qualifiedName,
+						metaEntity = AnnotationMetaEntity.create( element, context )
+				);
```

Comment:
I don't understand why you want to use `Metamodel`s here. They're just for code generation. (The name is very misleading.) And calling `addMetaEntity()` from here seems like it would have weird side-effects on code generation.

To me, having the `ProcessorSessionFactory` depend on the `Context` is a layer-breaker. I'm trying to keep a nice narrow interface between these two subsystems, and in particular, the query validator shouldn't depend on the code generator.

Reply:
I'm trying to reuse the type analysis part here, because I need to access the property types to discover enums. The fact that code generation is mixed into the `Metamodel` implementation is IMO a different problem that we should probably tackle separately.

Reply:
> The fact that code generation is mixed into the `Metamodel` implementation

Nono, you're misunderstanding. That's not a problem. That's fine and I have no plans to change it. (I go plan to change the _names_ though.)

What I'm saying is that the package `org.hibernate.processor.validation` should have no dependence on the code generation stuff in `org.hibernate.processor`, `org.hibernate.processor.annotation`, etc. It's supposed to be completely self-contained.

Reply:
If `ProcessorSessionFactory` needs to cache property types, and if we're sure it's truly worth the memory cost of doing so, then I guess it could cache them internally (lazily) from the `propertyType()` method.

Reply:
And, FTR, I don't like the way the old code builds up all these `Metamodel` objects and then writes them out. I don't really understand why it wasn't written to just write out the `java` code as soon as we finish building the `Metamodel` object, so that that object it can be immediately garbage collected. And I will likely change that, which would probably break the code you've written here. That's one reason why we should not break the layering here.